### PR TITLE
[clang] Fix crash when #embed used in a compound literal

### LINKED ISSUE #102248: `#embed` crashes when used in a compound literal
This valid C23 source file:

```c
char *p = (char[]){
#embed "/usr/bin/true"
};
```

[crashes clang](https://godbolt.org/z/9q5Wevsvo) somewhere in `InitListChecker::HandleEmbed`.

### PR DESCRIPTION
**Pull request by Fznamznon**
Fixes https://github.com/llvm/llvm-project/issues/102248

### CODE REVIEW

File: clang/test/Preprocessor/embed_codegen.cpp [hunk 1]
```diff
@@ -1,9 +1,15 @@
...
+// CHECK: @.compoundliteral = internal global [2 x i8] c"jk"
 // CHECK: @__const._Z3fooi.ca = private unnamed_addr constant [3 x i32] [i32 0, i32 106, i32 107], align 4
 // CHECK: @__const._Z3fooi.sc = private unnamed_addr constant %struct.S1 { i32 106, i32 107, i32 0 }, align 4
 // CHECK: @__const._Z3fooi.t = private unnamed_addr constant [3 x %struct.T] [%struct.T { [2 x i32] [i32 48, i32 49], %struct.S1 { i32 50, i32 51, i32 52 } }, %struct.T { [2 x i32] [i32 53, i32 54], %struct.S1 { i32 55, i32 56, i32 57 } }, %struct.T { [2 x i32] [i32 10, i32 0], %struct.S1 zeroinitializer }], align 16
 // CHECK: @__const._Z3fooi.W = private unnamed_addr constant %struct.Wrapper { i32 48, %struct.HasCharArray { [10 x i8] c"123456789\0A" } }, align 4
+
+char *p = (char[]){
+#embed "jk.txt"
+};
+
```
Comment (cor3ntin) [inline clang/test/Preprocessor/embed_codegen.cpp:12]: Can you add tests for `char[N]`, `int[]` ?

**Review by cor3ntin — COMMENTED** [review]
LGTM.
I think it is reasonable to backport that (in which case we don't need the changelog entry)

**Review by AaronBallman — APPROVED** [review]
Changes LGTM as far as they go, but some additional test coverage is requested.

File: clang/test/Preprocessor/embed_codegen.cpp (hunk 1 above)
Comment (Fznamznon) [reply clang/test/Preprocessor/embed_codegen.cpp]: No problem. Is it ok now?

**Review by AaronBallman — APPROVED** [review]
LGTM!

**Pull request merged**